include/RevCPU.h

```diff
@@ -212,6 +221,63 @@ class RevCPU : public SST::Component{
   RevLoader *Loader;                  ///< RevCPU: RISC-V loader
   std::vector<RevProc *> Procs;       ///< RevCPU: RISC-V processor objects
   bool *Enabled;                      ///< RevCPU: Completion structure
+      
+  // Global table of threads
+  // - The actual RevThread objects live in this table 
+  // - Only place where the presence of a RevThread on this CPU is guanranteed throughout execution
+  std::map<uint32_t, std::shared_ptr<RevThread>> Threads;
+
+  // 2D vector to hold threads assigned to each processor and its corresponding harts.
+  // AssignedThreads[i][j] holds the RevThread executing on processor 'i' and hart 'j'.
+  // Oversubscription is not supported in hardware 
+  // (ie. AssignedThreads.at(i).size() will never exceed the number of HARTS)
+  std::vector<std::vector<std::shared_ptr<RevThread>>> AssignedThreads;
+
+  // Initializes a RevThread object.
+  // - Moves it to the 'Threads' map 
+  // - Adds it's ThreadID to the ThreadQueue to be scheduled
+  void InitThread(std::shared_ptr<RevThread> ThreadToInit);
+
+  // Adds Thread with ThreadID to AssignedThreads vector for ProcID 
+  // - Handles updating LSQueue & MarkLoadComplete function pointers
+  void AssignThread(uint32_t ThreadID, uint32_t ProcID);
+
+  // Sets up arguments for a thread with a given ID and feature set.
+  void SetupArgs(uint32_t ThreadIDToSetup, RevFeature* feature);
+
+  // Checks the status of ALL threads that are currently blocked.
+  void CheckBlockedThreads();
+
+  // Checks core w/ ProcID to see if it has any available harts to assign work to
+  // if it does and there is work to assign (ie. ThreadQueue is not empty)
+  // assign it and enable the processor if not already enabled.
+  void UpdateThreadAssignments(uint32_t ProcID);
+
+  // Checks for state changes in the threads of a given processor index 'ProcID'
+  // and handle appropriately
+  void CheckForThreadStateChanges(uint32_t ProcID);
+
+  // Checks for new threads that may have been added to a given processor's NewThreadInfo 
+  void CheckForNewThreads(uint32_t ProcID);
+
+  // Checks if a thread with a given Thread ID can proceed (used for pthread_join).
+  // it does this by seeing if a given thread's WaitingOnTID has completed
+  bool ThreadCanProceed(uint32_t TID);
+
+  // Queue of Thread IDs waiting to be assigned. The actual RevThread objects reside in the 'Threads' map.
+  std::vector<uint32_t> ThreadQueue = {};
+
+  // Set of Thread IDs that are currently blocked (waiting for their WaitingOnTID to be present in CompletedThreads).
+  std::set<uint32_t> BlockedThreads = {};
+
+  // Set of Thread IDs that have completed their execution.
+  std::set<uint32_t> CompletedThreads = {};
+
+  // Random Number Generator for generating new Thread IDs.
+  SST::RNG::MarsagliaRNG* RNG;
+
+  // Generates a new Thread ID using the RNG.
+  uint32_t GetNewThreadID() { return RNG->generateNextUInt32(); }
```

Comment:
See `include/RevRand.h`. Please do not use another RNG.

Reply:
If you need a `uint32_t` random value, use `RevRand(0, UINT32_MAX)`

Reply:
This was intentionally chosen because the SST RNG ensures no ThreadID collisions across the entire SST run. Using RevRand does not offer the same protection.

Reply:
SST has a standard randomization function that allows us to generate pseudorandom numbers that are unique across MPI ranks and/or CXX threads.  In order to support forthcoming scalable multithreaded system architecture simulation, we need to ensure thread ID uniqueness across instances of Rev.  Using the SST rand functions are the easiest way to do so

Reply:
> Using RevRand does not offer the same protection.

This can be fixed.

The same argument could be used to say that `RevRand`, and the old `rand()` code it replaced, could not ensure **k-distrbutivity** across multiple Rev instances (that's the technical term for what you mean when you say "no collisions"), in other parts of the code besides the thread code. The thread code is not special in this regard -- what you're asking for could be asked for by all other uses of `RevRand`.

Reply:
> SST has a standard randomization function that allows us to generate pseudorandom numbers that are unique across MPI ranks and/or CXX threads. In order to support forthcoming scalable multithreaded system architecture simulation, we need to ensure thread ID uniqueness across instances of Rev. Using the SST rand functions are the easiest way to do so

Then `RevRand` needs to be reimplemented using those methods, which should be easy. I do not want a half-dozen RNGs everyone reinvents for their own area of code.

Reply:
I wonder why these issues were not brought up in the review of `RevRand()`. Sure, all of the previous uses of PRNGs were relatively trivial, like fault injection or simulating indeterminate memory latency, but the issue remains.

Reply:
I will immediately work on reimplementing `RevRand()` to use SST's RNGs. In the meantime, @rkabrick I stand by my suggestion that you should use the `RevRand()` API. It should be the central place where all PRNG features / requirements are met.

Reply:
Personal note: I spent 4 years as a hobby studying PRNGs in the late 1980s and early 1990s, and developed the PRNG in HP's Fortran 90 compiler, which was thread-safe and thread-specific but deterministically repeatable within each thread.

Reply:
I have a new version of `RevRand()` which uses SST's generator. No interface changes are necessary. I'll submit a PR shortly. This PR should use `RevRand()`.